### PR TITLE
KEYCLOAK-18368 Invalidate client session after refresh token re-use

### PR DESCRIPTION
**Pull request by mposolda**

### CODE REVIEW

File: server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java [hunk 1]
```diff
@@ -37,8 +37,16 @@
         public static final SearchableModelField<AuthenticatedClientSessionModel> TIMESTAMP  = new SearchableModelField<>("timestamp", Integer.class);
     }
 
+    String STARTED_AT_NOTE = "startedAt";
+
     String getId();
 
+    default int getStarted() {
```
Comment (pedroigor) [inline server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java:44]: Perhaps `getStartedAt` ?

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java [hunk 2]
```diff
@@ -403,14 +416,24 @@ private void validateTokenReuseForRefresh(KeycloakSession session, RealmModel re
...
+                validateTokenReuse(session, realm, refreshToken, clientSession, true);
                 int currentCount = clientSession.getCurrentRefreshTokenUseCount();
                 clientSession.setCurrentRefreshTokenUseCount(currentCount + 1);
+            } catch (OAuthErrorException oee) {
+                if (logger.isDebugEnabled()) {
+                    logger.debugf("Failed validation of refresh token %s due it was used before. Realm: %s, client: %s, user: %s, user session: %s. Will detach client session from user session",
+                            refreshToken.getId(), realm.getName(), clientSession.getClient().getClientId(), clientSession.getUserSession().getUser().getUsername(), clientSession.getUserSession().getId());
+                }
+                clientSession.detachFromUserSession();
+                new ResourceAdminManager(session).logoutClientSession(realm, clientSession.getClient(), clientSession);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:429]: Do we want to eventually send logout events to clients? Perhaps just detach the client session is enough?

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java [hunk 3]
```diff
@@ -174,6 +175,11 @@ public TokenValidation validateToken(KeycloakSession session, UriInfo uriInfo, C
             }
         }
 
+        if (oldToken.getIssuedAt() + 1 < clientSession.getStarted()) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:178]: This condition seems duplicated at different places. Does it make sense to overload `AccessToken.isActive` or another method to keep this logic in a single place?

File: server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java (hunk 1 above)
Comment (mposolda) [reply server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java:44]: Good point. I was thinking exactly about that and the reason why I did not is the fact UserSessionModel has method `getStarted()` . Not 100% sure whether to go with `getStartedAt`, which is probably slightly clearer name, or whether to stick with compatibility with UserSessionModel? Let me know your preference and I will do it that way.

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java (hunk 2 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:429]: Question is why not to send logout event? Any issue with that?

Usually when some possible attack is detected, we send an event to client (EG. when admin or user updates "notBefore" . Or when admin or user logouts through admin console or account console). The event is sent just in the case when client is using Keycloak server-side adapter and has "admin URL" configured (proprietary Keycloak mechanism).

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java (hunk 3 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:178]: Good point. I will try to extract possibly to some util method or something... Not sure about `AccessToken.isActive` as AuthenticatedClientSession can't be sent as an argument to AccessToken method (it is different module) .

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java (hunk 2 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:429]: You pretty much touched my point. We only send under specific circumstances. Make this possible also during validation is something I would avoid.

Depending on how many clients and how they are configured, you have an additional cost. With no real benefit as we have other means to advertise clients to invalidate tokens/sessions.

File: server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java (hunk 1 above)
Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java:44]: No preference now you gave your reasons. Up to you :)

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java (hunk 3 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:178]: Sure. We only need the int anyways.

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java (hunk 2 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java]: I am not convinced either way, but I am just removing it from this PR. From what I know, we might be more align to remove Keycloak proprietary "OIDC extensions" in the future anyway.

Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java]: Ok :) I keep the "started" due the compatibility with UserSessionModel.

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java (hunk 3 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java]: True. I've added method "isIssuedBeforeSessionStart" with the time as an argument. Also used this for few other places inside the codebase as there are similar existing checks for the userSession already existing.

File: server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java (hunk 1 above)
Comment (mposolda) [reply server-spi/src/main/java/org/keycloak/models/AuthenticatedClientSessionModel.java:44]: Ok :) I keep the "started" due the compatibility with UserSessionModel.

**Comment by mposolda** [comment]
@pedroigor Addressed your concerns. Do you please have a chance to re-review? Thanks

**Review by pedroigor — APPROVED** [review]

**Comment by mposolda** [comment]
@pedroigor Thanks!

**Pull request merged**